### PR TITLE
[Merged by Bors] - feat(RingTheory/Idempotents): Lifting complete orthogonal idempotents along nil extensions

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Idempotents.lean [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+The predicate `IsIdempotentElem` is defined for general monoids in `Algebra/Ring/Idempotents.lean`.
+In this file we provide various results regarding idempotent elements in rings.
+
+## Main definitions
+
+- `OrthogonalIdempotents`:
+  A family `{ eᵢ }` of idempotent elements is orthogonal if `eᵢ * eⱼ = 0` for all `i ≠ j`.
+- `CompleteOrthogonalIdempotents`:
+  A family `{ eᵢ }` of orthogonal idempotent elements is complete if `∑ eᵢ = 0`.
```
Comment (Julian-Kuelshammer) [inline Mathlib/RingTheory/Idempotents.lean:23]:
```suggestion
  A family `{ eᵢ }` of orthogonal idempotent elements is complete if `∑ eᵢ = 1`.
```

File: Mathlib/RingTheory/Idempotents.lean [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+
+## Main results
+
+- `CompleteOrthogonalIdempotents.lift_of_isNilpotent_ker`:
+  If the kernel of `f : R →+* S` consists of nilpotent elements, and `{ eᵢ }` is a family of
+  complete orthogonal idempotents in the range of `f`, then `{ eᵢ }` is the image of some
+  complete orthogonal idempotents in `R`.
+- `existsUnique_isIdempotentElem_eq_of_ker_isNilpotent`:
+  If `R` is commutative and the kernel of `f : R →+* S` consists of nilpotent elements,
+  then every idempotent in the range of `f` lifts to a unique idempotent in `S`.
```
Comment (Julian-Kuelshammer) [inline Mathlib/RingTheory/Idempotents.lean:33]:
```suggestion
  then every idempotent in the range of `f` lifts to a unique idempotent in `R`.
```

File: Mathlib/RingTheory/Idempotents.lean [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+- `CompleteOrthogonalIdempotents.bijective_pi`:
+  If `R` is commutative, then a family `{ eᵢ }` of complete orthogonal idempotent elements induces
+  an ring isomorphism `R ≃ ∏ R ⧸ ⟨1 - eᵢ⟩`.
```
Comment (Julian-Kuelshammer) [inline Mathlib/RingTheory/Idempotents.lean:36]:
```suggestion
  a ring isomorphism `R ≃ ∏ R ⧸ ⟨1 - eᵢ⟩`.
```

File: Mathlib/RingTheory/Idempotents.lean [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    {e : I → S} (he : CompleteOrthogonalIdempotents e) (he' : ∀ i, e i ∈ f.range) :
+    ∃ e' : I → R, CompleteOrthogonalIdempotents e' ∧ f ∘ e' = e := by
+  obtain ⟨e', h₁, h₂⟩ := lift_of_isNilpotent_ker_aux f h
+    ((equiv (Fintype.equivFin I).symm).mpr he) (fun _ ↦ he' _)
+  refine ⟨_, ((equiv (Fintype.equivFin I)).mpr h₁),
+    by ext x; simpa using congr_fun h₂ (Fintype.equivFin I x)⟩
+
+end Ring
+
+section CommRing
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/RingTheory/Idempotents.lean]: Thanks! I've generalised some lemmas by assuming centrality. Lemmas concerning quotients and products cannot be generalised though, and I've kept them as is.

**Comment by semorrison** [comment]
Looks good to me, but didn't look at proofs at all. I hope we can get this in soon, feel free to ping me if there's no movement and I'll hit merge!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**